### ISSUE TITLE
Composed Identities

### ISSUE THREAD

**Issue by jansentom** [objectiv/objectiv-analytics#961]
We've recently merged https://github.com/objectiv/objectiv-analytics/pull/936, which adds support for the IdentityContext, that contains an `id` and a `name`. Please have a look at the PR for more details. 

We'd like to be able to use these id/name pairs to "upgrade" the current cookie based `user_id` resolution something like this:
`mh.add_identity(name)` where `name` matches the name in the `IdentityContext` and the `id` is the field that makes it unique.

The session as currently defined by the session creation logic is used as the base for the new logic; when an identity is composed, it assigns the new, higher level identity to the session, thus stacking the logic. 

For the first version we'd like:
- The entire session is given to new identity
- The last new identity found in the session wins (this will be configurable in a later version)
- previous or later sessions by the same cookie_id, that do not yet contain the identity can optionally be assigned to the identity (think `ffill`, `bfill`) . `ffill` is the default behavior, but none or `bfill` should also be possible.
- all of this has to be specified before getting the objectiv dataframe.
- The `user_id` column contains the new unique identifier
- Old sessions that have no new identity and thus become anonymous get dropped, get NULL as their user_id, or they can keep their old user_id (conflict resolution for later versions) .
 
**Describe the solution/outcome you'd like**
A very simple interaction with mh should enrich the pipepline with identity composition. I'm thinking:
```
mh.add_identity('email', fill='both', anonymous='keep')
```
This would use any IdentityContexts with `name='email'`, assigning this new identity to all sessions where it occurs. All cookie_id based sessions that exist get selected, and because `fill='both'` all sessions with those those cookie_ids get assigned to the new identity. All anonymous sessions are kept with their original user_id.

**Comment by thijs-obj** [comment]
# IdentityContexts - `id` misused?
Are we not using `id` and `name` sort of in the reversed way from how we originally did it?

For other contexts the `id` has as a description: "A unique string identifier to be combined with the Context Type (`_type`) for Context instance uniqueness." [[1]](https://github.com/objectiv/objectiv-analytics/blob/main/schema/base_schema.json5#L139-L141)
So to me it would make most sense to use the `id` field for things like `name`, `email`, or `phone_number`, or `mobile_phone_number` or whatever the source of the identity, and not for the actual value. Because the `id` should be unique among the IdentityContexts, and that won't work now. 

## Example
Example: We have two pieces of identity information about a user, his phone number and his cell phone number. Currently we would model this as:
```json
[
    {"_type": "IdentityContext", "name": "phone", "id": "555-123456789"},
    {"_type": "IdentityContext", "name": "cell_phone", "id": "555-123456789"},
]
```
But this is not allowed! because we have two IdentityContexts with the same id. If we would model it like the following this problem would not exist:
```json
[
    {"_type": "IdentityContext", "id": "phone", "value": "555-123456789"},
    {"_type": "IdentityContext", "id": "cell_phone", "value": "555-123456789"},
]
```

[1]https://github.com/objectiv/objectiv-analytics/blob/main/schema/base_schema.json5#L139-L141

**Comment by thijs-obj** [comment]
## Some questions
> How this will show up in the dataframe
> 
> * The `user_id` column will contain the new unique identifier

Question: Do we not rely on this being a UUID in the modelhub?

> * The `session_id` column will be unique

Question: Why would the session_id be unique? I thought that in a standard Objectiv dataframe the unique field was event-id, and that there could be multiple events in a session?

**Comment by jansentom** [comment]
> Are we not using `id` and `name` sort of in the reversed way from how we originally did it?

Yes we are. Update the description; PR to update taxonomy will follow.

**Comment by jansentom** [comment]
> > How this will show up in the dataframe
> > 
> > * The `user_id` column will contain the new unique identifier
> 
> Question: Do we not rely on this being a UUID in the modelhub?

It just needs to be a unique string in its context. We do need to update some column definitions here and there.

**Comment by jansentom** [comment]
> > * The `session_id` column will be unique
> 
> Question: Why would the session_id be unique? I thought that in a standard Objectiv dataframe the unique field was event-id, and that there could be multiple events in a session?

That was phrases unclearly. Updated that. Sessions should remain intact and uniquely identifyable after user_ids get merged because of newly discovered identities.
